File: leetcode/go/1331.rank-transform-of-an-array.go
func arrayRankTransform(arr []int) []int {
    if len(arr) == 0 {
        return arr
    }
    orig := make([]int, len(arr))

    for i, n := range arr {
        orig[i] = n
    }

    sort.Ints(arr)

    ranks := make(map[int]int)
    ranks[arr[0]] = 1
    rank := 2

    for i := 1; i < len(arr); i++ {
        if arr[i] != arr[i-1] {
            ranks[arr[i]] = rank
            rank++
        }
    }

    ans := make([]int, len(arr))
    for i, n := range orig {
        ans[i] = ranks[n]
    }
    return ans
}

File: leetcode/python/1331.rank-transform-of-an-array.py
class Solution:
    def arrayRankTransform(self, arr: List[int]) -> List[int]:
        if not arr:
            return arr
        
        orig = [n for n in arr]

        arr.sort()

        ranks = {arr[0]: 1}
        rank = 2

        for i in range(1,len(arr)):
            if arr[i] != arr[i-1]:
                ranks[arr[i]] = rank
                rank += 1
        
        return [ranks[n] for n in orig]

